state/executor.go

```diff
@@ -123,6 +129,79 @@ type BlockResult struct {
 	TotalGas uint64
 }
 
+// getStakingEventType is a helper method for getting the type of system event
+func getStakingEventType(toAddress types.Address) staking.StakingEventType {
+	if toAddress == types.StringToAddress(system.StakingAddress) {
+		return staking.StakingEvent
+	}
+
+	return staking.UnstakingEvent
+}
+
+// constructPendingEvent is a helper method for constructing a pending event
+func constructPendingEvent(
+	toAddress types.Address,
+	fromAddress types.Address,
+	value *big.Int,
+	blockNumber uint64,
+) staking.PendingEvent {
+	eventType := getStakingEventType(toAddress)
+
+	return staking.PendingEvent{
+		Number:    int64(blockNumber),
+		Address:   fromAddress,
+		Value:     value,
+		EventType: eventType,
+	}
+}
+
+// cleanDiscardedEvents is a helper method for clearing discarded pending events
+func cleanDiscardedEvents(transactions []*types.Transaction, blockNumber uint64) {
+	for _, discardedTxn := range transactions {
+		// Remove a pending event for the discarded transaction
+		if system.IsSystemEvent(discardedTxn.To) {
+			pendingEvent := constructPendingEvent(
+				*discardedTxn.To,
+				discardedTxn.From,
+				discardedTxn.Value,
+				blockNumber,
+			)
+
+			hub := staking.GetStakingHub()
+
+			// Remove the event from the queue if it is present
+			hub.RemovePendingEvent(pendingEvent)
+		}
+	}
+}
+
+// commitApprovedEvents is a helper method for committing pending events that passed all checks
+func commitApprovedEvents(transactions []*types.Transaction, blockNumber uint64) {
+	for _, t := range transactions {
+		// Check if the transaction matches a staking / unstaking event
+		if system.IsSystemEvent(t.To) {
+			pendingEvent := constructPendingEvent(
+				*t.To,
+				t.From,
+				t.Value,
+				blockNumber,
+			)
+
+			hub := staking.GetStakingHub()
+
+			// Remove the event from the queue if it is present
+			if hub.RemovePendingEvent(pendingEvent) {
```

Comment:
Maybe log an error or warning if this returns false here?

Reply:
Got it. Logging an error if the event was not found 👍